core/src/poh_recorder.rs

```diff
@@ -176,14 +173,19 @@ impl PohRecorder {
     ) {
         self.clear_bank();
         let mut cache = vec![];
-        info!(
-            "reset poh from: {},{} to: {},{}",
-            self.poh.hash, self.poh.tick_height, blockhash, tick_height,
-        );
+        {
+            let mut poh = self.poh.lock().unwrap();
+            info!(
+                "reset poh from: {},{} to: {},{}",
+                poh.hash, self.tick_height, blockhash, tick_height,
+            );
+            poh.reset(blockhash, self.poh_config.hashes_per_tick);
+        }
+
         std::mem::swap(&mut cache, &mut self.tick_cache);
```

Comment:
I think you need to hold the `poh` lock before you drop the cache, otherwise somebody could call `tick()`, create a tick, and then we drop it here.  

So tick() and reset() used to be synchronized on `poh_recorder`, so you were guaranteed that changes to the internal `poh.recorder.poh` object wouldn't be couldn't be interleaved between tick() and reset(). Now I think there are potential races between tick() and reset(). 

@sagar-solana, @rob-solana, @pgarg66, I think we need to be pretty careful here...Can you guys spot any other issues?

Reply:
> So tick() and reset() used to be synchronized on poh_recorder, so you were guaranteed that changes to the internal poh.recorder.poh object wouldn't be couldn't be interleaved between tick() and reset()

I don't think this has changed.  The PohService will now hash like a wild man *only until* it's time to generate the tick hash or record an entry.  The actual `tick()` and `record()` are still fully synchronized within the main PohRecorder lock, as is `reset()` -- so there's still no way to generate a tick while `reset()` is executing

Reply:
It's confusing:
1. PohService must only call `Poh::hash()` which will never generate a tick.  But it may call `PohRecorder:tick()` to generate a tick, at which point it is synchronized with the rest of PohRecorder (but of course it must not hold the Poh lock when doing so, to avoid deadlocking)
2. PohRecorder must only call `Poh::tick()` or `Poh::record()` (and thus is synchronized with any callers of `PohRecorder::reset()`) .

Reply:
oh I see, when tick() is called, poh_recorder locks is held. Very subtle but cool.